### ISSUE TITLE
runtime: self-deadlock on mheap_.lock

### ISSUE THREAD

**Issue by mknyszek** [golang/go#64067]
```
#!watchflakes
default <- (log ~ `test timed out` || log ~ `Test killed with quit`) && (goarch == "ppc64" || goarch == "ppc64le")
```

After landing a bunch of changes today (and yesterday) the ppc64 builders started timing out:
- https://github.com/golang/go/issues/64062
- https://github.com/golang/go/issues/64050

The root cause of the issue is a self-deadlock on `mheap_.lock`. An invariant of this lock is that it must be held only on the system stack, because a stack growth may cause the lock to be acquired again when allocating a new stack.

This invariant isn't followed in a very particular function that was added last release. The fix is simple: follow the invariant.

What's bizarre is why this started showing up _now_, a very long time later. Another question is why it started showing up in response to a set of changes I landed today, and how those changes perturbed the runtime into triggering this case more often. This is also a potential problem on _all_ platforms, not just ppc64. I don't know what makes ppc64 more likely to trigger it.

One theory is that this call path is taken when the heap goal exceeds a certain threshold. The thing is, the heap goal is simply set arbitrarily high if `GOGC=off` (which is actually kind of a bug in the check, but that part is mostly harmless), so any test setting `GOGC=off` would be susceptible. I suspect it has something to do with this, but I am unsure.

I'm also unclear as to why this is happening in the compiler so often. Perhaps some packages we build create a heap big enough to exceed the check? Either that, or it could be related to the manipulation of GOGC that it does.

**Comment by bcmills** [comment]
Huh! Is there any chance this might also explain #56418?

**Comment by mknyszek** [comment]
Hm... I don't obviously see goroutines blocked on the same lock in https://github.com/golang/go/issues/56418, so I can't be sure. But two goroutines are suspiciously reported as "running on another thread" and they could in theory be blocked on this lock, so I can't rule it out though. I suppose we'll find out. :)

As an aside, I think we should have a checker mode for this invariant. In the vast majority of cases (like this one) it should be pretty obvious statically that we're acquiring this lock not on the system stack. The runtime does not make very many indirect calls.

**Comment by mknyszek** [comment]
@gopherbot Please open a backport issue for Go 1.21.

This issue can cause deadlocks in running programs. It's rare but there are no workarounds. It also only applies to code that landed in Go 1.21.
